File: .skills/openclaw-skills/skills/dadaniya99/xiaolongxia-feishu-card/scripts/send_card.py
#!/usr/bin/env python3
"""
飞书互动卡片发送脚本（国际版 Feishu schema 2.0 兼容）

用法：
  python3 send_card.py --open-id ou_xxx --title "标题" --content "**内容**"
  python3 send_card.py --open-id ou_xxx --title "标题" --content "内容" --template green
  python3 send_card.py --chat-id oc_xxx --title "标题" --content "内容"  # 发群聊

注意：content 支持 Markdown（**加粗**、*斜体*、`代码`、换行用 \\n）
"""

import argparse
import json
import os
import sys
import urllib.request
import urllib.error


APP_ID = "cli_a9f5877b3378dbd8"
OPENCLAW_CONFIG = os.path.expanduser("~/.openclaw/openclaw.json")


def get_app_secret():
    try:
        with open(OPENCLAW_CONFIG) as f:
            config = json.load(f)
        return config["channels"]["feishu"]["appSecret"]
    except Exception as e:
        print(f"❌ 读取 app_secret 失败: {e}", file=sys.stderr)
        sys.exit(1)


def get_token(app_id, app_secret):
    data = json.dumps({"app_id": app_id, "app_secret": app_secret}).encode()
    req = urllib.request.Request(
        "https://open.feishu.cn/open-apis/auth/v3/tenant_access_token/internal",
        data=data,
        headers={"Content-Type": "application/json"},
    )
    with urllib.request.urlopen(req, timeout=10) as resp:
        result = json.loads(resp.read())
    token = result.get("tenant_access_token")
    if not token:
        print(f"❌ 获取 token 失败: {result}", file=sys.stderr)
        sys.exit(1)
    return token


def send_card(token, receive_id, receive_id_type, title, content, template="blue"):
    card = {
        "schema": "2.0",
        "header": {
            "title": {"tag": "plain_text", "content": title},
            "template": template,
        },
        "body": {
            "elements": [
                {"tag": "markdown", "content": content}
            ]
        },
    }

    # 关键：双重 stringify
    payload = {
        "receive_id": receive_id,
        "msg_type": "interactive",
        "content": json.dumps(card),  # 第一次 stringify（card → string）
    }
    # 第二次 stringify 在 json.dumps(payload) 时自动发生（string → escaped string）

    data = json.dumps(payload).encode()
    req = urllib.request.Request(
        f"https://open.feishu.cn/open-apis/im/v1/messages?receive_id_type={receive_id_type}",
        data=data,
        headers={
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            result = json.loads(resp.read())
    except urllib.error.HTTPError as e:
        body = e.read().decode()
        print(f"❌ HTTP {e.code}: {body}", file=sys.stderr)
        sys.exit(1)

    if result.get("code") == 0:
        print(f"✅ 卡片发送成功！message_id: {result['data']['message_id']}")
    else:
        print(f"❌ 发送失败: code={result.get('code')} msg={result.get('msg')}", file=sys.stderr)
        sys.exit(1)


def main():
    parser = argparse.ArgumentParser(description="发送飞书互动卡片（schema 2.0）")
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("--open-id", help="接收者 open_id（个人 DM）")
    group.add_argument("--chat-id", help="群聊 chat_id")
    parser.add_argument("--title", required=True, help="卡片标题")
    parser.add_argument("--content", required=True, help="卡片正文（支持 Markdown）")
    parser.add_argument(
        "--template",
        default="blue",
        choices=["blue", "green", "red", "orange", "purple", "grey"],
        help="标题栏颜色（默认 blue）",
    )
    args = parser.parse_args()

    app_secret = get_app_secret()
    token = get_token(APP_ID, app_secret)

    if args.open_id:
        receive_id = args.open_id
        receive_id_type = "open_id"
    else:
        receive_id = args.chat_id
        receive_id_type = "chat_id"

    # 处理 \n 转义（shell 传入的字面 \n 转成真正换行）
    content = args.content.replace("\\n", "\n")

    send_card(token, receive_id, receive_id_type, args.title, content, args.template)


if __name__ == "__main__":
    main()

File: .skills/openclaw-skills/skills/hashstacs-hk/feishu-skills/feishu-auth/send-card.mjs
/**
 * feishu-auth/send-card.mjs — Public card sending module for all Feishu skills.
 *
 * Provides:
 *   - buildCard()    — Build Card 1.0 JSON from title/body/button params
 *   - sendCard()     — Send card to user with 3-level fallback (card → text → reply)
 *   - updateCard()   — PATCH-update an existing card message
 *
 * Usage (import):
 *   import { sendCard, updateCard } from '../feishu-auth/send-card.mjs';
 *   const result = await sendCard({ openId, body: '**Hello**', title: 'Title' });
 *
 * Usage (CLI):
 *   node send-card.mjs --open-id "ou_xxx" --body "**Hello**" [--title "Title"]
 *                      [--chat-id "oc_yyy"] [--button-text "Click"] [--button-url "https://..."]
 *                      [--color blue]
 *
 * All output is single-line JSON.
 */

import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { getConfig } from './token-utils.mjs';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// ---------------------------------------------------------------------------
// Low-level Feishu IM API
// ---------------------------------------------------------------------------

async function getTenantAccessToken(appId, appSecret) {
  const res = await fetch(
    'https://open.feishu.cn/open-apis/auth/v3/tenant_access_token/internal',
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ app_id: appId, app_secret: appSecret }),
    },
  );
  const json = await res.json();
  if (json.code !== 0) {
    throw new Error(`Failed to get tenant_access_token: ${json.msg}`);
  }
  return json.tenant_access_token;
}

async function sendFeishuMessage(tenantToken, receiveIdType, receiveId, cardContent, rawOverride) {
  const payload = rawOverride
    ? { receive_id: receiveId, ...rawOverride }
    : { receive_id: receiveId, msg_type: 'interactive', content: JSON.stringify(cardContent) };
  const res = await fetch(
    `https://open.feishu.cn/open-apis/im/v1/messages?receive_id_type=${receiveIdType}`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${tenantToken}`,
      },
      body: JSON.stringify(payload),
    },
  );
  return res.json();
}

async function updateFeishuMessage(tenantToken, messageId, cardContent) {
  const res = await fetch(
    `https://open.feishu.cn/open-apis/im/v1/messages/${messageId}`,
    {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${tenantToken}`,
      },
      body: JSON.stringify({ content: JSON.stringify(cardContent) }),
    },
  );
  return res.json();
}

// ---------------------------------------------------------------------------
// Card builder (Card 1.0 format — compatible with PATCH updates)
// ---------------------------------------------------------------------------

function buildCard({ title, body, buttonText, buttonUrl, color }) {
  const elements = [];
  if (body) elements.push({ tag: 'markdown', content: body });
  if (buttonUrl) {
    elements.push({
      tag: 'action',
      actions: [{
        tag: 'button',
        text: { tag: 'plain_text', content: buttonText || '点击打开' },
        type: 'primary',
        url: buttonUrl,
      }],
    });
  }
  const card = { config: { wide_screen_mode: true }, elements };
  if (title) {
    card.header = {
      title: { tag: 'plain_text', content: title },
      template: color || 'blue',
    };
  }
  return card;
}

// ---------------------------------------------------------------------------
// High-level: sendCard (build → send → fallback)
// ---------------------------------------------------------------------------

async function sendCard({
  openId,
  chatId,
  title,
  body,
  buttonText,
  buttonUrl,
  color,
  fallbackText,
}) {
  if (!openId) throw new Error('sendCard: openId is required');
  if (!body) throw new Error('sendCard: body is required');

  const cfg = getConfig(__dirname);
  const resolvedChatId = chatId || process.env.ENCLAWS_CHAT_ID || null;
  const receiveIdType = resolvedChatId ? 'chat_id' : 'open_id';
  const receiveId = resolvedChatId || openId;

  let tenantToken;
  try {
    tenantToken = await getTenantAccessToken(cfg.appId, cfg.appSecret);
  } catch (err) {
    return { success: false, error: err.message, reply: body + (buttonUrl ? `\n${buttonUrl}` : '') };
  }

  // Level 1: interactive card
  try {
    const card = buildCard({ title, body, buttonText, buttonUrl, color });
    const result = await sendFeishuMessage(tenantToken, receiveIdType, receiveId, card);
    if (result.code === 0) {
      return { success: true, message_id: result.data?.message_id };
    }
    process.stderr.write(`[send-card] card failed: code=${result.code} msg=${result.msg}\n`);
  } catch (err) {
    process.stderr.write(`[send-card] card error: ${err.message}\n`);
  }

  // Level 2: plain text fallback
  const text = fallbackText || (body + (buttonUrl ? `\n${buttonUrl}` : ''));
  try {
    const result = await sendFeishuMessage(tenantToken, receiveIdType, receiveId, null, {
      msg_type: 'text',
      content: JSON.stringify({ text }),
    });
    if (result.code === 0) {
      return { success: true, message_id: result.data?.message_id, fallback: true };
    }
    process.stderr.write(`[send-card] text fallback failed: code=${result.code} msg=${result.msg}\n`);
  } catch (err) {
    process.stderr.write(`[send-card] text fallback error: ${err.message}\n`);
  }

  // Level 3: return reply for Agent to display manually
  return { success: false, error: 'all send methods failed', reply: text };
}

// ---------------------------------------------------------------------------
// High-level: updateCard (build → PATCH)
// ---------------------------------------------------------------------------

async function updateCard({ messageId, title, body, buttonText, buttonUrl, color }) {
  if (!messageId) throw new Error('updateCard: messageId is required');
  if (!body) throw new Error('updateCard: body is required');

  const cfg = getConfig(__dirname);
  try {
    const tenantToken = await getTenantAccessToken(cfg.appId, cfg.appSecret);
    const card = buildCard({ title, body, buttonText, buttonUrl, color });
    const result = await updateFeishuMessage(tenantToken, messageId, card);
    if (result.code === 0) {
      return { success: true };
    }
    return { success: false, error: `code=${result.code} msg=${result.msg}` };
  } catch (err) {
    return { success: false, error: err.message };
  }
}

// ---------------------------------------------------------------------------
// Module exports
// ---------------------------------------------------------------------------

export {
  buildCard,
  sendCard,
  updateCard,
  getTenantAccessToken,
  sendFeishuMessage,
  updateFeishuMessage,
};

export default {
  buildCard,
  sendCard,
  updateCard,
  getTenantAccessToken,
  sendFeishuMessage,
  updateFeishuMessage,
};

// ---------------------------------------------------------------------------
// CLI mode (only when run directly: node send-card.mjs --open-id ...)
// ---------------------------------------------------------------------------

const isMainModule = process.argv[1] && import.meta.url.endsWith(process.argv[1].replace(/\\/g, '/').replace(/^[A-Z]:/, m => m.toLowerCase()));

if (isMainModule || process.argv[1]?.endsWith('send-card.mjs')) {
  const argv = process.argv.slice(2);
  function getArg(name) {
    const i = argv.indexOf(name);
    return i !== -1 && argv[i + 1] !== undefined ? argv[i + 1] : null;
  }

  const openId = getArg('--open-id');
  const body = getArg('--body');

  if (!openId || !body) {
    console.log(JSON.stringify({
      error: 'usage',
      message: 'Required: --open-id <id> --body <text>. Optional: --chat-id, --title, --button-text, --button-url, --color',
    }));
    process.exit(1);
  }

  sendCard({
    openId,
    chatId: getArg('--chat-id'),
    title: getArg('--title'),
    body,
    buttonText: getArg('--button-text'),
    buttonUrl: getArg('--button-url'),
    color: getArg('--color'),
    fallbackText: getArg('--fallback-text'),
  })
    .then(result => {
      console.log(JSON.stringify(result));
      process.exit(result.success ? 0 : 1);
    })
    .catch(err => {
      console.log(JSON.stringify({ success: false, error: err.message }));
      process.exit(1);
    });
}
